main.tf

```
@@ -22,6 +22,17 @@ resource "aws_s3_bucket" "this" {
       }
     }
   }
+
+  dynamic "grant" {
+    for_each = var.acl == null && length(var.grants) > 0 ? var.grants : []
```

This can really just be `var.grants`, no? The logic has the effect of giving `var.acl` precedence if it's set, but that's confusing behavior because it's a silent error if the user really wants the grant they set.

---

Yeah, suggest changing the default for `var.acl` to `null` (as the terraform/aws default will still use `private`), then simplifying this to just `var.grants`.
```suggestion
    for_each = var.grants
```